pkg/controller/autoimport/autoimport_controller.go

```diff
@@ -78,100 +77,233 @@ func (r *ReconcileAutoImport) Reconcile(ctx context.Context, request reconcile.R
 		return reconcile.Result{}, err
 	}
 
+	currentRetry := 0
+	if len(autoImportSecret.Annotations[constants.AnnotationAutoImportCurrentRetry]) != 0 {
+		currentRetry, err = strconv.Atoi(autoImportSecret.Annotations[constants.AnnotationAutoImportCurrentRetry])
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
 	reqLogger.Info("Reconciling auto import secret")
 
+	result, condition, rc, iErr := r.autoImport(ctx, managedClusterName, autoImportSecret)
+	if condition != nil {
+		if err := helpers.UpdateManagedClusterStatus(
+			r.client,
+			r.recorder,
+			managedClusterName,
+			*condition,
+		); err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+	if condition.Status == metav1.ConditionFalse && rc.current > currentRetry && rc.current < rc.total {
+		// update secret
+		if autoImportSecret.Annotations == nil {
+			autoImportSecret.Annotations = make(map[string]string)
+		}
+		autoImportSecret.Annotations[constants.AnnotationAutoImportCurrentRetry] = strconv.Itoa(rc.current)
+
+		if _, err := r.kubeClient.CoreV1().Secrets(managedClusterName).Update(
+			ctx, autoImportSecret, metav1.UpdateOptions{},
+		); err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	if iErr == nil {
+		if condition.Status == metav1.ConditionTrue {
+			reqLogger.Info(fmt.Sprintf("Managed cluster is imported, try to delete its auto import secret %s/%s",
+				autoImportSecret.Namespace, autoImportSecret.Name))
+			// delete secret
+			if err := helpers.DeleteAutoImportSecret(ctx, r.kubeClient, autoImportSecret, r.recorder); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+
+		return result, nil
+	}
+
+	if rc.current == rc.total {
+		r.recorder.Eventf("RetryTimesExceeded",
+			fmt.Sprintf("Exceed the retry times, delete the auto import secret %s/%s",
+				managedClusterName, autoImportSecret.Name))
+		// delete secret
+		if err := helpers.DeleteAutoImportSecret(ctx, r.kubeClient, autoImportSecret, r.recorder); err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+	return result, iErr
+}
+
+func newManagedClusterImportSucceededCondition(s metav1.ConditionStatus, reason, message string) *metav1.Condition {
+	return &metav1.Condition{
+		Type:    constants.ConditionManagedClusterImportSucceeded,
+		Status:  s,
+		Reason:  reason,
+		Message: message,
+	}
+}
+
+type importRetryCount struct {
+	total   int
+	current int
+}
+
+func (r *ReconcileAutoImport) autoImport(ctx context.Context, managedClusterName string,
+	autoImportSecret *v1.Secret) (reconcile.Result, *metav1.Condition, importRetryCount, error) {
+
+	reqLogger := log.WithValues("Request.Namespace", managedClusterName)
+
+	var err error
+	rc := importRetryCount{
+		total:   1,
+		current: 0,
+	}
+	if len(autoImportSecret.Data[constants.AutoImportRetryName]) != 0 {
```

Comment:
this is a good to validate the autoimport secret format firstly, we may have a validate function for this part to make sure the autoimport secret format is correct (also consider the hosted import)